### PR TITLE
linuxPackages.trelay: init at 22.03.5

### PR DESCRIPTION
**Pull request by aprilthepink**
###### Description of changes

Added kernel module trelay. 
Trelay is a kernel module that relays ethernet packets between two devices (similar to a bridge), but without any MAC address checks.
This makes it possible to bridge client mode or ad-hoc mode wifi devices to ethernet VLANs, assuming the remote end uses the same source MAC address as the device that packets are supposed to exit from.

Related to #241991

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by yayayayaka** [comment]
Result of `nixpkgs-review pr 243569` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>linuxKernel.packages.hardkernel_4_14.trelay</li>
    <li>linuxPackages_hardkernel_latest.trelay</li>
  </ul>
</details>
<details>
  <summary>23 packages built:</summary>
  <ul>
    <li>linuxKernel.packages.linux_4_14.trelay</li>
    <li>linuxPackages_4_14_hardened.trelay (linuxKernel.packages.linux_4_14_hardened.trelay)</li>
    <li>linuxKernel.packages.linux_4_19.trelay</li>
    <li>linuxPackages_4_19_hardened.trelay (linuxKernel.packages.linux_4_19_hardened.trelay)</li>
    <li>linuxKernel.packages.linux_5_10.trelay</li>
    <li>linuxPackages_5_10_hardened.trelay (linuxKernel.packages.linux_5_10_hardened.trelay)</li>
    <li>linuxKernel.packages.linux_5_15.trelay</li>
    <li>linuxPackages_5_15_hardened.trelay (linuxKernel.packages.linux_5_15_hardened.trelay)</li>
    <li>linuxKernel.packages.linux_5_4.trelay</li>
    <li>linuxPackages_5_4_hardened.trelay (linuxKernel.packages.linux_5_4_hardened.trelay)</li>
    <li>linuxPackages.trelay (linuxKernel.packages.linux_6_1.trelay)</li>
    <li>linuxPackages_hardened.trelay (linuxPackages_6_1_hardened.trelay)</li>
    <li>linuxKernel.packages.linux_6_3.trelay</li>
    <li>linuxPackages_6_3_hardened.trelay (linuxKernel.packages.linux_6_3_hardened.trelay)</li>
    <li>linuxPackages_latest.trelay (linuxKernel.packages.linux_6_4.trelay)</li>
    <li>linuxPackages_6_4_hardened.trelay (linuxKernel.packages.linux_6_4_hardened.trelay)</li>
    <li>linuxPackages_latest-libre.trelay (linuxKernel.packages.linux_latest_libre.trelay)</li>
    <li>linuxPackages-libre.trelay (linuxKernel.packages.linux_libre.trelay)</li>
    <li>linuxPackages_lqx.trelay (linuxKernel.packages.linux_lqx.trelay)</li>
    <li>linuxPackages_testing_bcachefs.trelay (linuxKernel.packages.linux_testing_bcachefs.trelay)</li>
    <li>linuxPackages_xanmod.trelay (linuxKernel.packages.linux_xanmod.trelay)</li>
    <li>linuxPackages_xanmod_latest.trelay (linuxKernel.packages.linux_xanmod_latest.trelay ,linuxPackages_xanmod_stable.trelay)</li>
    <li>linuxPackages_zen.trelay (linuxKernel.packages.linux_zen.trelay)</li>
  </ul>
</details>

**Review by yayayayaka — COMMENTED** [review]
Thanks for your first contribution to nixpkgs :heart:

The PR title and commit message could be renamed to `linuxPackages.trelay: init at 22.03.5`

**Comment by CutestNekoAqua** [comment]
What is hardkernel? I am well familiar with the hardened kernel (which the module builds for) but not with hardkernel. Google doesnt really help either.

File: pkgs/os-specific/linux/trelay/default.nix [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  makeFlags = [
+    "KERNELRELEASE=${kernel.modDirVersion}"
+    "KERNEL_DIR=${kernel.dev}/lib/modules/${kernel.modDirVersion}/build"
+    "INSTALL_MOD_PATH=$(out)"
+  ];
+
+  meta = with lib; {
+    description = "A kernel module that relays ethernet packets between two devices";
+    longDescription = ''
+      A kernel module that relays ethernet packets between two devices (similar to a bridge),
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/trelay/default.nix:33]: description and longDescription should not be duplicated

File: pkgs/os-specific/linux/trelay/default.nix [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    v = "22.03.5";
+in
+stdenv.mkDerivation rec {
+  name = "trelay";
+  version = "${v}-${kernel.version}";
+
+  src = fetchFromGitHub {
+    owner = "openwrt";
+    repo = "openwrt";
+    rev = "v${v}";
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/trelay/default.nix:12]:
```suggestion
{ stdenv, lib, fetchFromGitHub, kernel, kmod }:
let
  version = "22.03.5";
in
stdenv.mkDerivation {
  name = "trelay";
  version = "${version}-${kernel.version}";

  src = fetchFromGitHub {
    owner = "openwrt";
    repo = "openwrt";
    rev = "v${version}";
```
we should use sparseCheckout to not download the entire repo

File: pkgs/os-specific/linux/trelay/Makefile [hunk 3]
```diff
@@ -0,0 +1,14 @@
+KERNELRELEASE ?= $(shell uname -r)
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/trelay/Makefile:1]: We can't use https://github.com/openwrt/openwrt/blob/main/package/kernel/trelay/Makefile ?

**Review by SuperSandro2000 — COMMENTED** [review]
Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes) when naming your commits.

**Comment by SuperSandro2000** [comment]
> What is hardkernel? I am well familiar with the hardened kernel (which the module builds for) but not with hardkernel. Google doesnt really help either.

https://github.com/search?q=repo%3ANixOS%2Fnixpkgs%20hardkernel&type=code

just mark it broken for that

**Comment by CutestNekoAqua** [comment]
let me open this again in a new PR :woozy_face:

**Pull request closed without merging**